### PR TITLE
add tag controller

### PR DESCRIPTION
**Pull request by wy65701436**
use the tag controller to handle CRUD of tags, especially the delete scenario, it could validate
the immutable and signature. And move the code of tag handling from artifact controller to tag controller

Signed-off-by: wang yan <wangyan@vmware.com>

### CODE REVIEW

File: src/api/artifact/controller.go [hunk 1]
```diff
@@ -448,35 +407,25 @@ func (c *controller) copyDeeply(ctx context.Context, srcRepo, reference, dstRepo
 	return id, nil
 }
 
-func (c *controller) CreateTag(ctx context.Context, tag *Tag) (int64, error) {
+func (c *controller) CreateTag(ctx context.Context, tag *tag.Tag) (int64, error) {
```
Comment (heww) [inline src/api/artifact/controller.go]: Should this method move out of the artifact controller ?

File: src/api/artifact/controller.go [hunk 2]
```diff
@@ -448,35 +407,25 @@ func (c *controller) copyDeeply(ctx context.Context, srcRepo, reference, dstRepo
...
 	// TODO fire event
-	return c.tagMgr.Create(ctx, &(tag.Tag))
+	return c.tagCtl.Create(ctx, tag)
 }
-func (c *controller) ListTags(ctx context.Context, query *q.Query, option *TagOption) ([]*Tag, error) {
-	tgs, err := c.tagMgr.List(ctx, query)
+func (c *controller) ListTags(ctx context.Context, query *q.Query, option *tag.Option) ([]*tag.Tag, error) {
```
Comment (heww) [inline src/api/artifact/controller.go]: And move out this one?

File: src/api/artifact/controller.go [hunk 3]
```diff
@@ -448,35 +407,25 @@ func (c *controller) copyDeeply(ctx context.Context, srcRepo, reference, dstRepo
...
-	}
 	return tags, nil
 }
 
 func (c *controller) DeleteTag(ctx context.Context, tagID int64) error {
-	// Immutable checking is covered in middleware
-	// TODO check signature
 	// TODO fire delete tag event
-	return c.tagMgr.Delete(ctx, tagID)
+	return c.tagCtl.Delete(ctx, tagID)
```
Comment (heww) [inline src/api/artifact/controller.go]: Move out.

File: src/api/tag/controller.go [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+				WithMessage("the tag %s configured as immutable, cannot be updated", tag.Name)
+		}
+		if tag.Signed {
+			return ierror.New(nil).WithCode(ierror.PreconditionCode).
+				WithMessage("the tag %s with signature cannot be updated", tag.Name)
+		}
+		// the tag already exists under the repository and is attached to the artifact, return directly
+		if tag.ArtifactID == artifactID {
+			return nil
+		}
```
Comment (ywk253100) [inline src/api/tag/controller.go]: Put immutable and signature checking after this, if the tag already exists and is already attached to the same artifact, we shouldn't return error

File: src/api/tag/controller.go [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+	if err != nil {
+		return nil, err
+	}
+	tag.Tag = *daoTag
+
+	if option == nil {
+		return tag, nil
+	}
+
+	if option.WithImmutableStatus {
```
Comment (ywk253100) [inline src/api/tag/controller.go]: Call `assembleTag` directly

File: src/api/artifact/controller.go [hunk 6]
```diff
@@ -72,14 +69,14 @@ type Controller interface {
...
+	// Delete the artifact specified by artifact ID
 	Delete(ctx context.Context, id int64) (err error)
 	// Copy the artifact specified by "srcRepo" and "reference" into the repository specified by "dstRepo"
 	Copy(ctx context.Context, srcRepo, reference, dstRepo string) (id int64, err error)
 	// ListTags lists the tags according to the query, specify the properties returned with option
-	ListTags(ctx context.Context, query *q.Query, option *TagOption) (tags []*Tag, err error)
+	ListTags(ctx context.Context, query *q.Query, option *tag.Option) (tags []*tag.Tag, err error)
 	// CreateTag creates a tag
-	CreateTag(ctx context.Context, tag *Tag) (id int64, err error)
+	CreateTag(ctx context.Context, tag *tag.Tag) (id int64, err error)
```
Comment (ywk253100) [inline src/api/artifact/controller.go]: All tag related methods should be moved to tag controller

**Pull request merged**